sdks/python/apache_beam/transforms/combiners_test.py

```diff
@@ -392,5 +394,112 @@ def test_global_fanout(self):
       assert_that(result, equal_to([49.5]))
 
 
+class LatestTest(unittest.TestCase):
+
+  def test_globally(self):
+    l = [window.GlobalWindows.windowed_value(3, 100),
+         window.GlobalWindows.windowed_value(1, 200),
+         window.GlobalWindows.windowed_value(2, 300)]
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.Globally()
+      assert_that(latest, equal_to([2]))
+
+  def test_globally_empty(self):
+    l = []
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.Globally()
+      assert_that(latest, equal_to([None]))
+
+  def test_per_key(self):
+    l = [window.GlobalWindows.windowed_value(('a', 1), 300),
+         window.GlobalWindows.windowed_value(('b', 3), 100),
+         window.GlobalWindows.windowed_value(('a', 2), 200)]
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.PerKey()
+      assert_that(latest, equal_to([('a', 1), ('b', 3)]))
+
+  def test_per_key_empty(self):
+    l = []
+    with TestPipeline() as p:
+      pc = p | Create(l)
+      latest = pc | combine.Latest.PerKey()
+      assert_that(latest, equal_to([]))
+
+  def test_per_key_add_timestamp_element_type_violation(self):
+    l_int = [window.GlobalWindows.windowed_value(1, 300),
+             window.GlobalWindows.windowed_value(6, 900),
+             window.GlobalWindows.windowed_value(8, 500)]
+    l_dict = [window.GlobalWindows.windowed_value({'a': 5}, 200),
+              window.GlobalWindows.windowed_value({'b': 3}, 300),
+              window.GlobalWindows.windowed_value({'a': 8}, 100)]
+    l_3_tuple = [window.GlobalWindows.windowed_value((1, 2, 3), 800),
+                 window.GlobalWindows.windowed_value((4, 5, 6), 200),
+                 window.GlobalWindows.windowed_value((7, 8, 9), 700)]
+
+    with self.assertRaises(TypeCheckError):
+      with TestPipeline() as p:
+        pc = p | Create(l_int)
+        _ = pc | beam.ParDo(combine.Latest.PerKey.add_timestamp)
+
+    with self.assertRaises(TypeCheckError):
```

Comment:
What is being tested here? Why passing KV values here will result in an assert?

Reply:
This was to test that the `_check_instance_type` used above raises an Exception. The last positional argument to `window.GlobalWindows.windowed_value` is the timestamp of each element. So, the input PCollections are those that violate the KV constraint.